skyllh/core/analysis_utils.py

```diff
@@ -83,8 +87,88 @@ def calculate_pval_from_trials(
 
     return (p, p_sigma)
 
-def polynomial_fit(
-        ns, p, p_weight, deg, p_thr):
+def truncated_gamma_logpdf(
+        a, scale, eta, ts_above_eta, N_above_eta):
+    """Calculates the -log(likelihood) of a sample of random numbers
+    generated from a gamma pdf truncated from below at x=eta.
+
+    Parameters
+    ----------
+    a : float
+        Shape parameter.
+    scale : float
+        Scale parameter.
+    eta : float
+        Test-statistic value at which the gamma function is truncated
+        from below.
+    ts_above_eta : (n_trials,)-shaped 1D ndarray
+        The ndarray holding the test-statistic values falling in
+        the truncated gamma pdf.
+    N_above_eta : int
+        Number of test-statistic values falling in the truncated
+        gamma pdf.
+        
+    Returns
+    -------
+    -logl : float
+    """
+    c0 = 1. - gamma.cdf(eta, a=a, scale=scale)
+    c0 = 1./c0
+    logl = N_above_eta*np.log(c0) + np.sum(gamma.logpdf(ts_above_eta,
+                                                        a=a, scale=scale))
+    return -logl
+
+def calculate_critical_ts_from_gamma(
+        ts, h0_ts_quantile, eta=3.0, xi=1.e-2):
+    """Calculates the critical test-statistic value corresponding
+    to h0_ts_quantile by fitting the ts distribution with a truncated
+    gamma function.
+
+    Parameters
+    ----------
+    ts : (n_trials,)-shaped 1D ndarray
+        The ndarray holding the test-statistic values of the trials.
+    h0_ts_quantile : float
+        Null-hypothesis test statistic quantile.
+    eta : float, optional
+        Test-statistic value at which the gamma function is truncated
+        from below.
+    xi : float, optional
+        Approximate estimate of the median of the test-statistic
+        distribution.
+
+    Returns
+    -------
+    critical_ts : float
+    """
+    Ntot = len(ts)
+    N = len(ts[ts > xi])
+    alpha = N/Ntot
+
+    ts_eta = ts[ts > eta]
+    N_prime = len(ts_eta)
+    alpha_prime = N_prime/N
+
+    obj = lambda x: truncated_gamma_logpdf(x[0], x[1], eta=eta,
+                                           ts_above_eta=ts_eta,
+                                           N_above_eta=N_prime)
+    x0 = [0.75, 1.8]  # Initial values of function parameters.
+    bounds = [[0.1, 10], [0.1, 10]]  # Ranges for the minimization fitter.
+    r = minimize(obj, x0, bounds=bounds)
+    pars = r.x
+    
+    norm = alpha*(alpha_prime/(1 - gamma.cdf(eta, a=pars[0], scale=pars[1])))
```

Comment:
Survival function (also defined as 1 - cdf, but sf is sometimes more accurate).
```suggestion
    norm = alpha*(alpha_prime/gamma.sf(eta, a=pars[0], scale=pars[1]))
```